lib/cn/csched_sp3.c

```diff
@@ -1954,118 +1972,134 @@ sp3_schedule(struct sp3 *sp)
+             */
+            if (sp->lpct_targ < rp_leaf_pct)
+                job = sp3_check_rb_tree(sp, RBT_RI_ALEN, 0, wtype_ispill, qnum);
+            break;
+
+        case jtype_node_len:
+            qnum = SP3_QNUM_NODELEN;
+            if (qfull(sp, qnum) && shared_full)
                 break;
 
-            case jtype_node_len:
-                /* Service RBT_LI_LEN red-black tree.
-                 * Implements:
-                 *   - Internal node query-shape rule
-                 *   - Leaf node query-shape rule
-                 */
-                qi = sp->qinfo + SP3_QNUM_INTERN;
-                if (qfull(qi) && shared_full)
-                    break;
-                job = sp3_check_rb_tree(sp, RBT_LI_LEN, node_len_max, wtype_node_len);
+            /* Service RBT_LI_LEN red-black tree.
+             * Implements:
+             *   - Internal node query-shape rule
+             *   - Leaf node query-shape rule
+             */
+            job = sp3_check_rb_tree(sp, RBT_LI_LEN, node_len_max, wtype_node_len, qnum);
+            break;
+
+        case jtype_leaf_garbage:
+            qnum = SP3_QNUM_LEAF;
+            if (qfull(sp, qnum))
                 break;
 
-            case jtype_leaf_garbage:
-                /* Service RBT_L_GARB red-black tree.
-                 * Implements:
-                 *   - Leaf node space amp rule
-                 * Notes:
-                 *   - These are big jobs, so do not use shared
-                 *     workers to limit the number of concurrent
-                 *     big jobs to the number dedicated workers.
-                 *   - Don't check for garbage unless ucomp is active
-                 *     or if in samp_reduce mode and leaf percent is
-                 *     somewhat caught up (ie, current leaf pct
-                 *     (lpct_targ) is within 90% of rparam setting
-                 *     (rp_leaf_pct)).
-                 *   - When checking for garbage, if leaf percent is
-                 *     behind, then bump up threshold so we don't waste
-                 *     write amp by compacting nodes with a small
-                 *     amount of garbage (we'd rather wait for
-                 *     leaf_pct to catch up).
-                 */
-                qi = sp->qinfo + SP3_QNUM_LEAF;
-                if (qfull(qi))
-                    break;
-                if (sp->samp_reduce && (100 * sp->lpct_targ > 90 * rp_leaf_pct)) {
-                    uint thresh = sp->lpct_targ < rp_leaf_pct ? 10 : 0;
-                    job = sp3_check_rb_tree(sp, RBT_L_GARB, thresh, wtype_leaf_garbage);
-                }
+            /* Service RBT_L_GARB red-black tree.
+             * Implements:
+             *   - Leaf node space amp rule
+             * Notes:
+             *   - These are big jobs, so do not use shared
+             *     workers to limit the number of concurrent
+             *     big jobs to the number dedicated workers.
```

Comment:
Comment says shared queue is not used, but I don't see a check for `shared_full` here.